### PR TITLE
Model/melgan

### PR DESCRIPTION
**Pull request by Ardiegon**
Implementation of MelGAN.
Some of formating, tensorboard and typing is little bit off, I will split it and upload in another PR. In this one i'm mainly uploading functionalities of training and inference.
IMPORTANT! Changed librosa version to 0.9.1, yell in comments if something is broken by this change.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -194,3 +194,6 @@ venv.bak/
 # dev experiments
 tests/experiments
 local_workspace/
+
+# local data dirs
+data/
```
Comment (Szakulli07) [inline .gitignore:199]: new line at the end

File: coopertunes/datasets/AudioDataset.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    """
+    This is the main class that returns audio frames
+    """
+
+    def __init__(self, training_files, segment_length, sampling_rate, augment=True):
+        self.sampling_rate = sampling_rate
+        self.segment_length = segment_length
+        self.audio_files = files_to_list(training_files)
+        # self.audio_files = [ x for x in self.audio_files]
+        random.seed(1234)
```
Comment (Szakulli07) [inline coopertunes/datasets/AudioDataset.py:34]: there is func in utils for setting all seeds `sed_seed`

File: coopertunes/datasets/AudioDataset.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    files = [f.rstrip() for f in files]
+    return files
+
+
+class AudioDataset(torch.utils.data.Dataset):
+    """
+    This is the main class that returns audio frames
+    """
+
+    def __init__(self, training_files, segment_length, sampling_rate, augment=True):
```
Comment (Szakulli07) [inline coopertunes/datasets/AudioDataset.py:29]: please move it to hparams

File: coopertunes/hparams/MelGan.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        self.batch_size= 64
+        self.downsamp_factor= 4
+        self.epochs= 3000
+        self.lambda_feat= 10
+        self.log_interval= 100
+        self.n_test_samples= 8
+        self.save_interval= 1000
+        self.load_path= ""
+        self.data_path= ""
+        self.save_path= "summary"
```
Comment (Szakulli07) [inline coopertunes/hparams/MelGan.py:30]: if possible please use params same as in base `HParams`

File: coopertunes/hparams/MelGan.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        self.ndf= 16
+        self.n_layers_D= 4
+        self.cond_disc= False
+        self.num_D= 3
+        self.n_mel_channels= 80
+        self.batch_size= 64
+        self.downsamp_factor= 4
+        self.epochs= 3000
+        self.lambda_feat= 10
+        self.log_interval= 100
```
Comment (Szakulli07) [inline coopertunes/hparams/MelGan.py:25]: same as above

File: coopertunes/hparams/Audio2Mel.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+from pathlib import Path
+from typing import Any, Optional, Union
+
+from coopertunes.hparams import HParams
+
+
+class Audio2MelHParams(HParams):
```
Comment (Szakulli07) [inline coopertunes/hparams/Audio2Mel.py:7]: Generally if you need it only for MelGan it could be in MelGan hparams

File: coopertunes/hparams/MelSpecVAE.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
 from pathlib import Path
 from typing import Any, Optional, Union
 
-from .hparams import HParams
+from coopertunes.hparams import HParams
```
Comment (Szakulli07) [inline coopertunes/hparams/MelSpecVAE.py:4]: Okay, why we don't like relative imports?

File: coopertunes/hparams/hparams.py [hunk 8]
```diff
@@ -71,3 +72,6 @@ def _repr_object(self, v):
 
     def __repr__(self) -> str:
         return self._dumps()
+    
+    def get(self,name):
+        return vars(self)[name][0]
```
Comment (Szakulli07) [inline coopertunes/hparams/hparams.py:77]: <3

File: coopertunes/models/Audio2Mel.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import torch.nn as nn
+import torch.nn.functional as F
+import torch
+from librosa.filters import mel as librosa_mel_fn
+
+from coopertunes.hparams import HParams
+
+
+class Audio2Mel(nn.Module):
```
Comment (Szakulli07) [inline coopertunes/models/Audio2Mel.py:9]: O wow, why we need a `nn.Module` for that. if you need mels there is also func for it in utils

Comment (Szakulli07) [reply coopertunes/models/Audio2Mel.py:9]: If you need differentiable mels there is still `torchaudio.transforms.MelSpectrogram`

File: coopertunes/models/MelGan.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        self.shortcut = WNConv1d(dim, dim, kernel_size=1)
+
+    def forward(self, x):
+        return self.shortcut(x) + self.block(x)
+
+
+class MelGanGenerator(nn.Module):
+    def __init__(self, hparams: HParams):
+        super().__init__()
+        ratios = [8, 8, 2, 2]
```
Comment (Szakulli07) [inline coopertunes/models/MelGan.py:44]: mayby it can also be in hparams

File: coopertunes/models/MelGan.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            nn.LeakyReLU(0.2),
+            WNConv1d(dim, dim, kernel_size=1),
+        )
+        self.shortcut = WNConv1d(dim, dim, kernel_size=1)
+
+    def forward(self, x):
+        return self.shortcut(x) + self.block(x)
+
+
+class MelGanGenerator(nn.Module):
```
Comment (Szakulli07) [inline coopertunes/models/MelGan.py:41]: Please add inference if possible. I mean for usage after training there is no need for returning feature map

File: coopertunes/supervisors/Audio2Mel.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+from coopertunes.models import Audio2Mel
```
Comment (Szakulli07) [inline coopertunes/supervisors/Audio2Mel.py:1]: wait, w want to train converting Audio2Mels?

File: coopertunes/supervisors/MelGan.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+        print(f"Chosen device: {self.device}")
+
+        self.epoch = 1
+        self.step = 1
+
+        self._logger = Logger("melgan", self.hparams, device)
+        self.writer = self._logger.get_summary_writer()
+
+        self.optG = torch.optim.Adam(self.netG.parameters(), lr=1e-4, betas=(0.5, 0.9))
+        self.optD = torch.optim.Adam(self.netD.parameters(), lr=1e-4, betas=(0.5, 0.9))
```
Comment (Szakulli07) [inline coopertunes/supervisors/MelGan.py:36]: totally hparamsable

File: coopertunes/utils.py [hunk 14]
```diff
@@ -196,6 +204,18 @@ def calc_n_params(module):
...
 
+def save_sample(file_path, sampling_rate, audio):
+    """Helper function to save sample
+
+    Args:
+        file_path (str or pathlib.Path): save file path
+        sampling_rate (int): sampling rate of audio (usually 22050)
+        audio (torch.FloatTensor): torch array containing audio in [-1, 1]
+    """
+    audio = (audio.numpy() * 32768).astype("int16")
```
Comment (Szakulli07) [inline coopertunes/utils.py:215]: ` * 32768?`

File: requirements.txt [hunk 15]
```diff
@@ -0,0 +1,4 @@
+librosa==0.9.1
+scipy==1.11.4
+argparse
+tensorboard
```
Comment (Szakulli07) [inline requirements.txt:4]: please move those to pyproject.toml

File: README.md [hunk 16]
```diff
@@ -25,6 +26,8 @@ Before install `coopertunes` module you need to install `pytorch` framework. It
 Now you need to install `coopertunes` module:
 
 `pip install -e .`
+`pip install -r requirements.txt`
```
Comment (Szakulli07) [inline README.md:29]: please use only pyproject.toml

File: requirements.txt (hunk 15 above)
Comment (Szakulli07) [reply requirements.txt:4]: ans specify version of argparse. Tensorboard is available in torch

**Review by Szakulli07 — APPROVED** [review]
Well if this works than why not, i will leave there some random thoughts.

File: coopertunes/hparams/Audio2Mel.py (hunk 6 above)
Comment (Ardiegon) [reply coopertunes/hparams/Audio2Mel.py:7]: It could be used for any other training when online spectrogram calculation  could be benefitial (adding noise masks, random crops etc.)

File: coopertunes/datasets/AudioDataset.py (hunk 3 above)
Comment (Ardiegon) [reply coopertunes/datasets/AudioDataset.py:29]: Already in hparams, imo at this level it should be steered by top level abstraction (here model supervisors with its own hparams)

File: coopertunes/hparams/MelSpecVAE.py (hunk 7 above)
Comment (Ardiegon) [reply coopertunes/hparams/MelSpecVAE.py:4]: Without using specific IDE, not using relative imports could be more readable to python interpreter. Specific imports are more robust. For example, working with project by setting PYTHONPATH to project root, could provoke errors sometimes.

File: coopertunes/hparams/hparams.py (hunk 8 above)
Comment (Ardiegon) [reply coopertunes/hparams/hparams.py:77]: YEEEE ITS NEEDED FOR ALL OF US TO FORGOR IT FASTTTTTTT

Some backstory: I acidentally saved myself one element tuples in Hparams attributes, and debbuged it for a while why tf it's automatically changing attributes to tuples (is it some weird feature of hparam?). It's a relict of a very frustrating half an hour

File: coopertunes/models/Audio2Mel.py (hunk 9 above)
Comment (Ardiegon) [reply coopertunes/models/Audio2Mel.py:9]: Sure, this one is for processing audio online if there is a need (explanation in some other comment here above). Generally as we implement it as module, we can run it in series with trained model on GPU.

File: coopertunes/models/MelGan.py (hunk 11 above)
Comment (Ardiegon) [reply coopertunes/models/MelGan.py:41]: I see in MelSpec Vae inference is just the decoding part of architecture. I believe in case of MelGAN eval inference and training inference looks the same

File: coopertunes/supervisors/Audio2Mel.py (hunk 12 above)
Comment (Ardiegon) [reply coopertunes/supervisors/Audio2Mel.py:1]: Yes, explanation above.

File: coopertunes/utils.py (hunk 14 above)
Comment (Ardiegon) [reply coopertunes/utils.py:215]: Not exactly sure what it was too, but now it is named macro ;)

**Comment by Ardiegon** [comment]
As it comes to checks for linting: I'm waiting for automisation of code formatting, dodging extra rewriting of code.
